certora/specs/DoubleEndedQueue.spec

```diff
@@ -0,0 +1,366 @@
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: pushBack adds an element at the end of the queue                                              │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBack(bytes32 value) {
+    require boundedQueue();
+
+    uint256 lengthBefore = length();
+    
+    pushBack@withrevert(value);
+    
+    // liveness
+    assert !lastReverted, "never reverts";
+    
+    // effect
+    assert back() == value, "back set to value";
+    assert length() == lengthBefore + 1, "queue increased";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: pushBack preserves the previous values in the queue                                                           │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule pushBackConsistency(uint256 key) {
+    require boundedQueue();
+
+    bytes32 beforeAt = at_(key);
+
+    bytes32 value;
+    pushBack(value);
+
+    // try to read value
+    bytes32 afterAt = at_@withrevert(key);
+
+    assert !lastReverted, "value still there";
+    assert afterAt == beforeAt, "data is preserved";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Function correctness: popFront removes an element from the beginning of the queue                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFront {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    uint256 lengthBefore = length();
+    bytes32 frontBefore = front@withrevert();
+
+    bytes32 popped = popFront@withrevert();
+    bool success = !lastReverted;
+
+    // liveness
+    assert success <=> lengthBefore != 0, "never reverts if not previously empty";
+
+    // effect
+    assert success => frontBefore == popped, "previous front is returned";
+    assert success => length() == lengthBefore - 1, "queue decreased";
+}
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: at(x) is preserved and offset to at(x - 1) after calling popFront                                             |
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule popFrontConsistency(uint256 key) {
+    requireInvariant boundariesConsistency();
+    require boundedQueue();
+
+    // Read (any) value that is not the front (this asserts the value exists / the queue is long enough)
+    require key > 1;
+    bytes32 before = at_(key);
```

Comment:
I just realized that here (and almost everywhere else) you use `key` ... but this is more of an `index`. 

Looking at DoubleEndedQueue alone its not obvious, but when considering EnumerableSet and EnumerableMap ... I think we should be consistent and use `index`

Reply:
Done :D